pkg/trace/api/listener.go

```diff
@@ -77,12 +78,16 @@ func (ln *measuredListener) flushMetrics() {
 
 type onCloseConn struct {
 	net.Conn
-	onClose func()
+	closeOnce sync.Once
+	onClose   func()
 }
 
 func (c *onCloseConn) Close() error {
-	err := c.Conn.Close()
-	c.onClose()
+	var err error
+	c.closeOnce.Do(func() {
+		err = c.Conn.Close()
+		c.onClose()
+	})
 	return err
 }
```

Comment:
Great 👍 If we can add a quick unit test for this that'd be great, but if it's too much of a pain, then no worries.